ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,500 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+ *
+ * @param[in] Amount
+ * The amount of quota to be returned.
+ *
+ * @return
+ * Nothing.
  */
 VOID
 NTAPI
-PspReturnProcessQuotaSpecifiedPool(IN PEPROCESS Process,
-                                   IN PS_QUOTA_TYPE QuotaType,
-                                   IN SIZE_T    Amount)
+PspReturnProcessQuotaSpecifiedPool(
+    _In_opt_ PEPROCESS Process,
+    _In_ PEPROCESS_QUOTA_BLOCK QuotaBlock,
+    _In_ PS_QUOTA_TYPE QuotaType,
+    _In_ SIZE_T Amount)
 {
     KIRQL OldIrql;
-    ASSERT(Process);
-    ASSERT(Process != PsInitialSystemProcess);
+    SIZE_T ReturnThreshold;
+    SIZE_T AmountToReturn = 0;
+    SIZE_T UpdatedQuotaUsage = 0;
+
+    /* Sanity checks */
     ASSERT(QuotaType < PsQuotaTypes);
-    ASSERT(!(Amount & 0x80000000)); /* we need to be able to negate it */
+    ASSERT((SSIZE_T)Amount >= 0);
 
-    /* Guard our quota in a spin lock */
+    /* Guard ourselves in a spin lock */
     KeAcquireSpinLock(&PspQuotaLock, &OldIrql);
 
-    if (Process->QuotaUsage[QuotaType] < Amount)
+    /* Does the caller return more quota than it was previously charged? */
+    if ((Process && Process->QuotaUsage[QuotaType] < Amount) ||
+        QuotaBlock->QuotaEntry[QuotaType].Usage < Amount)
     {
-        DPRINT1("WARNING: Process->QuotaUsage sanity check failed.\n");
+        /* It does, crash the system! */
+        KeBugCheckEx(QUOTA_UNDERFLOW,
+                     (ULONG_PTR)Process,
+                     (ULONG_PTR)QuotaType,
+                     Process ? (ULONG_PTR)Process->QuotaUsage[QuotaType] :
+                               QuotaBlock->QuotaEntry[QuotaType].Usage,
+                     (ULONG_PTR)Amount);
     }
-    else
+
+    /*
+     * We need to trim the quota limits based on the
+     * amount we're going to return quotas back.
+     */
+    if (QuotaType != PsPageFile &&
+        QuotaBlock != &PspDefaultQuotaBlock)
+    {
+        /* The return threshold can be non paged or paged */
+        ReturnThreshold = QuotaType ? PSP_NON_PAGED_POOL_QUOTA_THRESHOLD : PSP_PAGED_POOL_QUOTA_THRESHOLD;
+
+        /*
+         * If the amount to return exceeds the threshold,
+         * the new amount becomes the default, otherwise
+         * the amount is just the one given by the caller.
+         */
+        AmountToReturn = min(Amount, ReturnThreshold);
+
+        /* Return the amounting we have back to Memory Manager */
+        InterlockedExchangeAddSizeT(QuotaBlock->QuotaEntry[QuotaType].Return, AmountToReturn);
+        MmReturnPoolQuota(QuotaType, QuotaBlock->QuotaEntry[QuotaType].Return);
+
+        /* And try to trim the limit */
+        InterlockedExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Limit,
+                                 QuotaBlock->QuotaEntry[QuotaType].Limit - AmountToReturn);
+    }
+
+    /* Update the usage member of the block */
+    UpdatedQuotaUsage = QuotaBlock->QuotaEntry[QuotaType].Usage - Amount;
+    InterlockedExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Usage, UpdatedQuotaUsage);
```

Comment:
Nit: Right now this is correct (cause we use the spin lock everywhere), but I'd recommend also making this an `ExchangeAdd`, just like for `QuotaUsage` below, because that's both simpler (IMHO) and will remain correct once we remove the lock (no double-fetch)